### PR TITLE
Fix type annotation

### PR DESCRIPTION
**Pull request by egor43**
For issue #254 
unittest.patch return MagicMock or AsyncMock: https://github.com/python/cpython/blob/3.11/Lib/unittest/mock.py#L1467

### CODE REVIEW

**Comment by egor43** [comment]
@nicoddemus Do you have time to take a look?

**Comment by The-Compiler** [comment]
Looking at [what typeshed does](https://github.com/python/typeshed/blob/main/stdlib/unittest/mock.pyi#L275-L278), this seems to b a step in the right direction - though it breaks things on Python 3.7.

I suppose we could drop 3.7 support now, rather than introducing a conditional (given that [it's EOL](https://devguide.python.org/versions/) in 19 days)?

**Comment by egor43** [comment]
I can do the same:
```python
if sys.version_info >= (3, 8):
    _Mock: TypeAlias = MagicMock | AsyncMock
else:
    _Mock: TypeAlias = MagicMock
```
Would such a solution work?

**Comment by egor43** [comment]
Thank you @nicoddemus.
Sorry I didn't quite understand about using UNRELEASED for version. Could you tell me more?

**Comment by nicoddemus** [comment]
The changes for each version starts with a header, for example `3.10.0 (2022-10-05)`, but the next version is not released yet, so just use `UNRELEASED`.

**Comment by egor43** [comment]
I updated the changelog and made changes to the type annotation. Could you review the changes?

File: src/pytest_mock/plugin.py [hunk 1]
```diff
@@ -112,7 +118,7 @@ def stop(self, mock: unittest.mock.MagicMock) -> None:
         else:
             raise ValueError("This mock object is not registered")
 
-    def spy(self, obj: object, name: str) -> unittest.mock.MagicMock:
+    def spy(self, obj: object, name: str) -> MockType:
```
Comment (nicoddemus) [inline src/pytest_mock/plugin.py:121]: Thanks @egor43!

It seems to me though, by returning a Union, that will force users to check the return type before using it:

```python
foo_mock = mocker.spy(module, some_async_foo)
assert isinstance(foo_mock, mocker.AsyncMock)
# use foo_mock as mocker.AsyncMock.
```

If the above is right, I don't think we need this PR, because that should work regardless -- however I think it will make mypy consider the code below it unreachable...

Possibly a better way to type annotate this would be to use `@overload`s, with different return types based on the input object, something like this:

https://github.com/python/typeshed/blob/852ac3efb3da704267082cc611557214f2f87aac/stubs/mock/mock/mock.pyi#L244-L270C17

As before thanks for the PR, but I wanted to discuss this to ensure we are getting the correct fix.

Comment (nicoddemus) [reply src/pytest_mock/plugin.py:121]: Btw using `@overloads` here seems very complicated, given the really dynamic nature of `mock` -- I was not asking you to do it,  I was just commenting on a possibility, and see if I might be missing something.

Comment (nicoddemus) [reply src/pytest_mock/plugin.py:121]: Did some testing:

```python
import unittest.mock
from pytest_mock import MockerFixture

class Foo:
    def normal_func(self) -> None:
        ...

    async def async_func(self) -> None:
        ...

async def test_x(mocker: MockerFixture) -> None:
    m = mocker.patch.object(Foo, "normal_func")
    reveal_type(m)
    await m()
    m.assert_called()
    assert m.await_count == 1

async def test_y() -> None:
    with unittest.mock.patch.object(Foo, "normal_func") as m:
        reveal_type(m)
        await m()
        m.assert_called()
        assert m.await_count == 1
```

Running `mypy` in `main`:

```
λ mypy .tmp\test_foo.py
.tmp\test_foo.py:13: note: Revealed type is "unittest.mock.MagicMock"
.tmp\test_foo.py:20: note: Revealed type is "Union[unittest.mock.MagicMock, unittest.mock.AsyncMock]"
Success: no issues found in 1 source file
```

So right of the gate I'm surprised here: I expected `mypy` to complain in `test_x` about being awaited and `m.await_count`, given those attribute are from `AsyncMock` and do not exist in `MagicMock`. However of course this is the nature of `mock`, it will automatically create any attribute when accessed.

Running this on your branch:

```
λ mypy .tmp\test_foo.py
.tmp\test_foo.py:13: note: Revealed type is "Union[unittest.mock.MagicMock, unittest.mock.AsyncMock, unittest.mock.NonCallableMagicMock]"
.tmp\test_foo.py:20: note: Revealed type is "Union[unittest.mock.MagicMock, unittest.mock.AsyncMock]"
Success: no issues found in 1 source file
```

So the type annotation in `test_x` is now "correct" -- I use "correct" in quotes because as far as I understand, `mypy` cannot really help here, because it does not matter if `mocker.patch` returns `MagicMock` or `MagicMock | AsyncMock`, because it does not matter what attribute is accessed, it must always succeed.

I wonder what was failing in #254.

-----

So if my understand is correct, unfortunately the new type annotations does not seem to be helpful at all in catching errors.

Having said all that, I think it is harmless to merge the PR, at least we will be better align with how `mock` is typed.

Comment (nicoddemus) [reply src/pytest_mock/plugin.py:121]: I've asked @GDYendell for clarification in #254.

Comment (egor43) [reply src/pytest_mock/plugin.py:121]: Thanks for the in-depth analysis of the issue and for the flaw found.
Yes, I was thinking about using ```@overloads``` . But it really seemed to me a very difficult decision. In addition, ```unittest.mock._patch``` does a lot of runtime checks to infer the final type.

Comment (egor43) [reply src/pytest_mock/plugin.py:121]: You're right. mypy does not react in any way to an attempt to access non-existent attributes.
But some IDEs like VSCode highlight the attribute as unknown. Because of this, you cannot jump to the implementation of the method.
My current pull request returns a more truthy type that both mypy and the IDE can work with.

Comment (nicoddemus) [reply src/pytest_mock/plugin.py:121]: Thanks @GDYendell, as per information you provided in #254, I managed to reproduce the issue you had using this code:

```python
from unittest.mock import AsyncMock

from pytest_mock import MockerFixture


class Control:

    async def _start_processes(self) -> None:
        ...

    async def start_processes(self, x: int) -> None:
        ...


async def test_start_processes_press(
        control: Control, mocker: MockerFixture
) -> None:
    mocker.patch.object(control, "_start_processes", new_callable=AsyncMock)
    await control.start_processes(1)
    control._start_processes.assert_awaited_once_with()
```

This gives the following errors, both in this PR and `main`:

```
 λ mypy .tmp\test_foo2.py
.tmp\test_foo2.py:20: error: "Callable[[], Coroutine[Any, Any, None]]" has no attribute "assert_awaited_once_with"
[attr-defined]
        control._start_processes.assert_awaited_once_with()
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 1 source file)
```

Which is the same error you had when you reported #254.

But the problem here is how it is written: `mypy` can't know that `control._start_processes` has been mocked and is now a `MagicMock`, so it correctly points the problem that a callable does not have the `assert_awaited_once_with` method.

The fix is to simply use the `MagicMock` object returned by `mocker.patch.object`:

```python
async def test_start_processes_press(
        control: Control, mocker: MockerFixture
) -> None:
    m = mocker.patch.object(control, "_start_processes", new_callable=AsyncMock)
    await control.start_processes(1)
    m.assert_awaited_once_with()
```

So this passes, again both in this PR and in `main`.

It was fixed in that repository in https://github.com/dls-controls/odinprocservcontrol/commit/6056b44c96b0f946ebb07039e096782a888f8798 by doing essentially the same thing, but using `unittest.mock` directly:

```python
async def test_start_processes_press(
        control: Control
) -> None:
    with unittest.mock.patch.object(control, "_start_processes", new_callable=AsyncMock) as m:
        await control.start_processes(1)
        m.assert_awaited_once_with()
```

In conclusion, the problem was never the type annotation. :sweat_smile:

I think however given that this PR does improve IDE auto-completion, I see no harm merging it.

Thanks everyone!

**Review by nicoddemus — APPROVED** [review]
